drivers/bus/isapnp/isapnp.c

```diff
@@ -194,323 +152,295 @@ IsaConvertMemRange32Requirement(
-            for (i = 0; i < Alternatives->Count; i++)
+            case ISAPNP_RESOURCE_TYPE_IRQ:
+            case ISAPNP_RESOURCE_TYPE_DMA:
             {
-                TempBuffer = Alternatives->Irq[i].Mask;
-                RtlInitializeBitMap(&TempBitmap,
-                                    &TempBuffer,
-                                    RTL_BITS_OF(Alternatives->Irq[i].Mask));
-                BitCount = RtlNumberOfSetBits(&TempBitmap);
-
-                if (BitCount > 1)
-                    AltOptionalCount += BitCount - 1;
+                RTL_BITMAP ResourceBitmap;
+                ULONG BitmapSize, BitmapBuffer, BitCount;
+
+                if (Resource->Type == ISAPNP_RESOURCE_TYPE_IRQ)
+                {
+                    BitmapSize = RTL_BITS_OF(Resource->IrqDescription.Mask);
+                    BitmapBuffer = Resource->IrqDescription.Mask;
+                }
+                else
+                {
+                    BitmapSize = RTL_BITS_OF(Resource->DmaDescription.Mask);
+                    BitmapBuffer = Resource->DmaDescription.Mask;
+                }
+                RtlInitializeBitMap(&ResourceBitmap, &BitmapBuffer, BitmapSize);
+
+                BitCount = RtlNumberOfSetBits(&ResourceBitmap);
+                switch (DfState)
+                {
+                    case dfNotStarted:
+                        FirstFixedDescriptors += BitCount;
+                        break;
+
+                    case dfStarted:
+                        ResourceCount += BitCount;
+                        break;
+
+                    case dfDone:
+                        LastFixedDescriptors += BitCount;
+                        break;
+
+                    DEFAULT_UNREACHABLE;
+                }
+
+                break;
             }
-        }
-        if (HasDmaAlternatives(Alternatives))
-        {
-            for (i = 0; i < Alternatives->Count; i++)
+
+            case ISAPNP_RESOURCE_TYPE_IO:
+            case ISAPNP_RESOURCE_TYPE_MEMRANGE:
+            case ISAPNP_RESOURCE_TYPE_MEMRANGE32:
             {
-                TempBuffer = Alternatives->Dma[i].Mask;
-                RtlInitializeBitMap(&TempBitmap,
-                                    &TempBuffer,
-                                    RTL_BITS_OF(Alternatives->Dma[i].Mask));
-                BitCount = RtlNumberOfSetBits(&TempBitmap);
-
-                if (BitCount > 1)
-                    AltOptionalCount += BitCount - 1;
+                switch (DfState)
+                {
+                    case dfNotStarted:
+                        ++FirstFixedDescriptors;
+                        break;
+
+                    case dfStarted:
+                        ++ResourceCount;
+                        break;
+
+                    case dfDone:
+                        ++LastFixedDescriptors;
+                        break;
+
+                    DEFAULT_UNREACHABLE;
+                }
+                break;
             }
+
+            default:
```

Comment:
Not a `DEFAULT_UNREACHABLE` as well?

Reply:
No `DEFAULT_UNREACHABLE` is intended here

Reply:
why? (it looks like a default_unreachable to me :thinking: )

Reply:
Updated to make it more clear.

Reply:
sorry, where is the update or explanation? (you could tell us here in the PR comments)

Reply:
I added `UNREACHABLE` there and a case for the `ISAPNP_RESOURCE_TYPE_END_DEPENDENT` with an empty code block.